Tests failing

Tests are failing

`go test ./...` results
```
# github.com/tynany/frr_exporter/collector [github.com/tynany/frr_exporter/collector.test]
collector/bgp_test.go:277:29: not enough arguments in call to processBGPSummary
	have (chan prometheus.Metric, []byte, string)
	want (chan<- prometheus.Metric, []byte, string, string)
collector/bgp_test.go:280:29: not enough arguments in call to processBGPSummary
	have (chan prometheus.Metric, []byte, string)
	want (chan<- prometheus.Metric, []byte, string, string)
?   	github.com/tynany/frr_exporter	[no test files]
FAIL	github.com/tynany/frr_exporter/collector [build failed]
FAIL
```

I am working on a fix.

--- a/collector/bgp.go
+++ b/collector/bgp.go
@@ -15,10 +15,12 @@ import (
 )
 
 var (
-	bgpSubsystem        = "bgp"
-	bgpPeerMetricPrefix = "bgp_peer"
+	bgpSubsystem         = "bgp"
+	bgpPeerMetricPrefix  = "bgp_peer"
+	bgpL2vpnMetricPrefix = "bgp_l2vpn_evpn"
 
-	bgpDesc = map[string]*prometheus.Desc{}
+	bgpDesc      map[string]*prometheus.Desc
+	bgpL2vpnDesc map[string]*prometheus.Desc
 
 	bgpErrors           = []error{}
 	totalBGPErrors      = 0.0
@@ -58,10 +60,7 @@ func (*BGPCollector) EnabledByDefault() bool {
 
 // Describe implemented as per the prometheus.Collector interface.
 func (*BGPCollector) Describe(ch chan<- *prometheus.Desc) {
-	if len(bgpDesc) == 0 {
-		setDesc()
-	}
-	for _, desc := range bgpDesc {
+	for _, desc := range getBgpDesc() {
 		ch <- desc
 	}
 }
@@ -106,10 +105,7 @@ func (*BGP6Collector) EnabledByDefault() bool {
 
 // Describe implemented as per the prometheus.Collector interface.
 func (*BGP6Collector) Describe(ch chan<- *prometheus.Desc) {
-	if len(bgpDesc) == 0 {
-		setDesc()
-	}
-	for _, desc := range bgpDesc {
+	for _, desc := range getBgpDesc() {
 		ch <- desc
 	}
 }
@@ -154,17 +150,69 @@ func (*BGPL2VPNCollector) EnabledByDefault() bool {
 
 // Describe implemented as per the prometheus.Collector interface.
 func (*BGPL2VPNCollector) Describe(ch chan<- *prometheus.Desc) {
-	if len(bgpDesc) == 0 {
-		setDesc()
-	}
-	for _, desc := range bgpDesc {
+	for _, desc := range getBgpDesc() {
 		ch <- desc
 	}
+	for _, desc := range getBgpL2vpnDesc() {
+		ch <- desc
+	}
+}
+
+func getBgpL2vpnEvpnSummary() ([]byte, error) {
+	return execVtyshCommand("-c", "show evpn vni json")
+}
+
+type vxLanStats struct {
+	Vni            int
+	VxlanType      string `json:"type"`
+	VxlanIf        string
+	NumMacs        float64
+	NumArpNd       float64
+	NumRemoteVteps float64
+	TenantVrf      string
+}
+
+func execVtyshCommand(args ...string) ([]byte, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), vtyshTimeout)
+	defer cancel()
+
+	output, err := exec.CommandContext(ctx, vtyshPath, args...).Output()
+	if err != nil {
+		return nil, err
+	}
+	return output, nil
+}
+
+func processBgpL2vpnEvpnSummary(ch chan<- prometheus.Metric, jsonBGPL2vpnEvpnSum []byte) error {
+	var jsonMap map[string]vxLanStats
+	bgpL2vpnDesc := getBgpL2vpnDesc()
+	if err := json.Unmarshal(jsonBGPL2vpnEvpnSum, &jsonMap); err != nil {
+		return fmt.Errorf("cannot unmarshal outputs of 'show evpn vni json': %s", err)
+	}
+
+	for _, vxLanStat := range jsonMap {
+		bgpL2vpnLabels := []string{strconv.Itoa(vxLanStat.Vni), vxLanStat.VxlanType, vxLanStat.VxlanIf, vxLanStat.TenantVrf}
+		newGauge(ch, bgpL2vpnDesc["numMacs"], vxLanStat.NumMacs, bgpL2vpnLabels...)
+		newGauge(ch, bgpL2vpnDesc["numArpNd"], vxLanStat.NumArpNd, bgpL2vpnLabels...)
+		newGauge(ch, bgpL2vpnDesc["numRemoteVteps"], vxLanStat.NumRemoteVteps, bgpL2vpnLabels...)
+	}
+	return nil
 }
 
 // Collect implemented as per the prometheus.Collector interface.
 func (c *BGPL2VPNCollector) Collect(ch chan<- prometheus.Metric) {
 	collectBGP(ch, "l2vpn")
+
+	jsonBGPL2vpnEvpnSum, err := getBgpL2vpnEvpnSummary()
+	if err != nil {
+		totalBGPL2VPNErrors++
+		bgpL2VPNErrors = append(bgpL2VPNErrors, fmt.Errorf("cannot execute 'show evpn vni json': %s", err))
+	} else {
+		if err := processBgpL2vpnEvpnSummary(ch, jsonBGPL2vpnEvpnSum); err != nil {
+			totalBGPL2VPNErrors++
+			bgpL2VPNErrors = append(bgpL2VPNErrors, err)
+		}
+	}
 }
 
 // CollectErrors returns what errors have been gathered.
@@ -177,7 +225,11 @@ func (*BGPL2VPNCollector) CollectTotalErrors() float64 {
 	return totalBGPL2VPNErrors
 }
 
-func setDesc() {
+func getBgpDesc() map[string]*prometheus.Desc {
+	if bgpDesc != nil {
+		return bgpDesc
+	}
+
 	bgpLabels := []string{"vrf", "afi", "safi", "local_as"}
 	bgpPeerTypeLabels := []string{"type", "afi", "safi"}
 	bgpPeerLabels := append(bgpLabels, "peer", "peer_as")
@@ -202,6 +254,21 @@ func setDesc() {
 		"UptimeSec":             colPromDesc(bgpPeerMetricPrefix, "uptime_seconds", "How long has the peer been up.", bgpPeerLabels),
 		"peerTypesUp":           colPromDesc(bgpPeerMetricPrefix, "types_up", "Total Number of Peer Types that are Up.", bgpPeerTypeLabels),
 	}
+
+	return bgpDesc
+}
+
+func getBgpL2vpnDesc() map[string]*prometheus.Desc {
+	if bgpL2vpnDesc != nil {
+		return bgpL2vpnDesc
+	}
+	bgpL2vpnLabels := []string{"vni", "type", "vxlanIf", "tenantVrf"}
+	bgpL2vpnDesc = map[string]*prometheus.Desc{
+		"numMacs":        colPromDesc(bgpL2vpnMetricPrefix, "num_macs_count_total", "Number of known MAC addresses", bgpL2vpnLabels),
+		"numArpNd":       colPromDesc(bgpL2vpnMetricPrefix, "num_arp_nd_count_total", "Number of ARP / ND entries", bgpL2vpnLabels),
+		"numRemoteVteps": colPromDesc(bgpL2vpnMetricPrefix, "num_remote_vteps_count_total", "Number of known remote VTEPs", bgpL2vpnLabels),
+	}
+	return bgpL2vpnDesc
 }
 
 func collectBGP(ch chan<- prometheus.Metric, AFI string) {
@@ -249,19 +316,12 @@ func collectBGP(ch chan<- prometheus.Metric, AFI string) {
 func getBGPSummary(AFI string, SAFI string) ([]byte, error) {
 	args := []string{"-c", fmt.Sprintf("show bgp vrf all %s %s summary json", AFI, SAFI)}
 
-	ctx, cancel := context.WithTimeout(context.Background(), vtyshTimeout)
-	defer cancel()
-
-	output, err := exec.CommandContext(ctx, vtyshPath, args...).Output()
-	if err != nil {
-		return nil, err
-	}
-	return output, nil
+	return execVtyshCommand(args...)
 }
 
 func processBGPSummary(ch chan<- prometheus.Metric, jsonBGPSum []byte, AFI string, SAFI string) error {
 	var jsonMap map[string]bgpProcess
-
+	bgpDesc := getBgpDesc()
 	if err := json.Unmarshal(jsonBGPSum, &jsonMap); err != nil {
 		return fmt.Errorf("cannot unmarshal bgp summary json: %s", err)
 	}
@@ -284,7 +344,6 @@ func processBGPSummary(ch chan<- prometheus.Metric, jsonBGPSum []byte, AFI strin
 		procLabels := []string{strings.ToLower(vrfName), strings.ToLower(AFI), strings.ToLower(SAFI), localAs}
 		// No point collecting metrics if no peers configured.
 		if vrfData.PeerCount != 0 {
-
 			newGauge(ch, bgpDesc["ribCount"], vrfData.RIBCount, procLabels...)
 			newGauge(ch, bgpDesc["ribMemory"], vrfData.RIBMemory, procLabels...)
 			newGauge(ch, bgpDesc["peerCount"], vrfData.PeerCount, procLabels...)

--- a/collector/bgp_test.go
+++ b/collector/bgp_test.go
@@ -204,87 +204,111 @@ var (
 }
 }
 `)
+
+	evpnVniJson = []byte(`
+    {
+  "174374":{
+    "vni":174374,
+    "type":"L2",
+    "vxlanIf":"ONTEP1_174374",
+    "numMacs":42,
+    "numArpNd":0,
+    "numRemoteVteps":1,
+    "tenantVrf":"default",
+    "remoteVteps":[
+      "10.0.0.13"
+    ]
+  },
+  "172192":{
+    "vni":172192,
+    "type":"L2",
+    "vxlanIf":"ONTEP1_172192",
+    "numMacs":0,
+    "numArpNd":23,
+    "numRemoteVteps":1,
+    "tenantVrf":"default",
+    "remoteVteps":[
+      "10.0.0.13"
+    ]
+  }
+  }`)
 	expectedBGPMetrics = map[string]float64{
-		"frr_bgp_message_input_total{address_family=ipv4unicast,peer=192.168.0.2,vrf=default}":  100,
-		"frr_bgp_message_input_total{address_family=ipv4unicast,peer=192.168.0.3,vrf=default}":  0,
-		"frr_bgp_message_input_total{address_family=ipv4unicast,peer=192.168.1.2,vrf=red}":      100,
-		"frr_bgp_message_input_total{address_family=ipv4unicast,peer=192.168.1.3,vrf=red}":      200,
-		"frr_bgp_message_input_total{address_family=ipv6unicast,peer=fd00::1,vrf=default}":      29285,
-		"frr_bgp_message_input_total{address_family=ipv6unicast,peer=fd00::101,vrf=red}":        29285,
-		"frr_bgp_message_input_total{address_family=ipv6unicast,peer=fd00::105,vrf=red}":        0,
-		"frr_bgp_message_input_total{address_family=ipv6unicast,peer=fd00::5,vrf=default}":      0,
-		"frr_bgp_message_output_total{address_family=ipv4unicast,peer=192.168.0.2,vrf=default}": 100,
-		"frr_bgp_message_output_total{address_family=ipv4unicast,peer=192.168.0.3,vrf=default}": 0,
-		"frr_bgp_message_output_total{address_family=ipv4unicast,peer=192.168.1.2,vrf=red}":     100,
-		"frr_bgp_message_output_total{address_family=ipv4unicast,peer=192.168.1.3,vrf=red}":     200,
-		"frr_bgp_message_output_total{address_family=ipv6unicast,peer=fd00::1,vrf=default}":     29285,
-		"frr_bgp_message_output_total{address_family=ipv6unicast,peer=fd00::101,vrf=red}":       29285,
-		"frr_bgp_message_output_total{address_family=ipv6unicast,peer=fd00::105,vrf=red}":       0,
-		"frr_bgp_message_output_total{address_family=ipv6unicast,peer=fd00::5,vrf=default}":     0,
-		"frr_bgp_peer_groups_memory_bytes{address_family=ipv4unicast,vrf=default}":              0,
-		"frr_bgp_peer_groups_memory_bytes{address_family=ipv4unicast,vrf=red}":                  0,
-		"frr_bgp_peer_groups_memory_bytes{address_family=ipv6unicast,vrf=default}":              0,
-		"frr_bgp_peer_groups_memory_bytes{address_family=ipv6unicast,vrf=red}":                  0,
-		"frr_bgp_peer_groups{address_family=ipv4unicast,vrf=default}":                           0,
-		"frr_bgp_peer_groups{address_family=ipv4unicast,vrf=red}":                               0,
-		"frr_bgp_peer_groups{address_family=ipv6unicast,vrf=default}":                           0,
-		"frr_bgp_peer_groups{address_family=ipv6unicast,vrf=red}":                               0,
-		"frr_bgp_peer_up{address_family=ipv4unicast,peer=192.168.0.2,vrf=default}":              1,
-		"frr_bgp_peer_up{address_family=ipv4unicast,peer=192.168.0.3,vrf=default}":              0,
-		"frr_bgp_peer_up{address_family=ipv4unicast,peer=192.168.1.2,vrf=red}":                  1,
-		"frr_bgp_peer_up{address_family=ipv4unicast,peer=192.168.1.3,vrf=red}":                  0,
-		"frr_bgp_peer_up{address_family=ipv6unicast,peer=fd00::1,vrf=default}":                  1,
-		"frr_bgp_peer_up{address_family=ipv6unicast,peer=fd00::101,vrf=red}":                    1,
-		"frr_bgp_peer_up{address_family=ipv6unicast,peer=fd00::105,vrf=red}":                    0,
-		"frr_bgp_peer_up{address_family=ipv6unicast,peer=fd00::5,vrf=default}":                  0,
-		"frr_bgp_peer_uptime_seconds{address_family=ipv4unicast,peer=192.168.0.2,vrf=default}":  10,
-		"frr_bgp_peer_uptime_seconds{address_family=ipv4unicast,peer=192.168.0.3,vrf=default}":  0,
-		"frr_bgp_peer_uptime_seconds{address_family=ipv4unicast,peer=192.168.1.2,vrf=red}":      20,
-		"frr_bgp_peer_uptime_seconds{address_family=ipv4unicast,peer=192.168.1.3,vrf=red}":      0,
-		"frr_bgp_peer_uptime_seconds{address_family=ipv6unicast,peer=fd00::1,vrf=default}":      87873,
-		"frr_bgp_peer_uptime_seconds{address_family=ipv6unicast,peer=fd00::101,vrf=red}":        87873,
-		"frr_bgp_peer_uptime_seconds{address_family=ipv6unicast,peer=fd00::105,vrf=red}":        0,
-		"frr_bgp_peer_uptime_seconds{address_family=ipv6unicast,peer=fd00::5,vrf=default}":      0,
-		"frr_bgp_peers_memory_usage_bytes{address_family=ipv4unicast,vrf=default}":              39936,
-		"frr_bgp_peers_memory_usage_bytes{address_family=ipv4unicast,vrf=red}":                  39936,
-		"frr_bgp_peers_memory_usage_bytes{address_family=ipv6unicast,vrf=default}":              59904,
-		"frr_bgp_peers_memory_usage_bytes{address_family=ipv6unicast,vrf=red}":                  59904,
-		"frr_bgp_peers{address_family=ipv4unicast,vrf=default}":                                 2,
-		"frr_bgp_peers{address_family=ipv4unicast,vrf=red}":                                     2,
-		"frr_bgp_peers{address_family=ipv6unicast,vrf=default}":                                 2,
-		"frr_bgp_peers{address_family=ipv6unicast,vrf=red}":                                     2,
-		"frr_bgp_prefixes_active{address_family=ipv4unicast,peer=192.168.0.2,vrf=default}":      0,
-		"frr_bgp_prefixes_active{address_family=ipv4unicast,peer=192.168.0.3,vrf=default}":      2,
-		"frr_bgp_prefixes_active{address_family=ipv4unicast,peer=192.168.1.2,vrf=red}":          2,
-		"frr_bgp_prefixes_active{address_family=ipv4unicast,peer=192.168.1.3,vrf=red}":          0,
-		"frr_bgp_prefixes_active{address_family=ipv6unicast,peer=fd00::1,vrf=default}":          1,
-		"frr_bgp_prefixes_active{address_family=ipv6unicast,peer=fd00::101,vrf=red}":            1,
-		"frr_bgp_prefixes_active{address_family=ipv6unicast,peer=fd00::105,vrf=red}":            0,
-		"frr_bgp_prefixes_active{address_family=ipv6unicast,peer=fd00::5,vrf=default}":          0,
-		"frr_bgp_rib_entries{address_family=ipv4unicast,vrf=default}":                           1,
-		"frr_bgp_rib_entries{address_family=ipv4unicast,vrf=red}":                               0,
-		"frr_bgp_rib_entries{address_family=ipv6unicast,vrf=default}":                           3,
-		"frr_bgp_rib_entries{address_family=ipv6unicast,vrf=red}":                               3,
-		"frr_bgp_rib_memory_usage_bytes{address_family=ipv4unicast,vrf=default}":                64,
-		"frr_bgp_rib_memory_usage_bytes{address_family=ipv4unicast,vrf=red}":                    0,
-		"frr_bgp_rib_memory_usage_bytes{address_family=ipv6unicast,vrf=default}":                456,
-		"frr_bgp_rib_memory_usage_bytes{address_family=ipv6unicast,vrf=red}":                    456,
+		"frr_bgp_peer_groups_count_total{afi=ipv4,local_as=64512,safi=unicast,vrf=default}":                                           0.0,
+		"frr_bgp_peer_groups_count_total{afi=ipv4,local_as=64612,safi=unicast,vrf=red}":                                               0.0,
+		"frr_bgp_peer_groups_count_total{afi=ipv6,local_as=64512,safi=unicast,vrf=default}":                                           0.0,
+		"frr_bgp_peer_groups_count_total{afi=ipv6,local_as=64612,safi=unicast,vrf=red}":                                               0.0,
+		"frr_bgp_peer_groups_memory_bytes{afi=ipv4,local_as=64512,safi=unicast,vrf=default}":                                          0.0,
+		"frr_bgp_peer_groups_memory_bytes{afi=ipv4,local_as=64612,safi=unicast,vrf=red}":                                              0.0,
+		"frr_bgp_peer_groups_memory_bytes{afi=ipv6,local_as=64512,safi=unicast,vrf=default}":                                          0.0,
+		"frr_bgp_peer_groups_memory_bytes{afi=ipv6,local_as=64612,safi=unicast,vrf=red}":                                              0.0,
+		"frr_bgp_peer_message_received_total{afi=ipv4,local_as=64512,peer=192.168.0.2,peer_as=64513,safi=unicast,vrf=default}":        100.0,
+		"frr_bgp_peer_message_received_total{afi=ipv4,local_as=64512,peer=192.168.0.3,peer_as=64514,safi=unicast,vrf=default}":        0.0,
+		"frr_bgp_peer_message_received_total{afi=ipv4,local_as=64612,peer=192.168.1.2,peer_as=64613,safi=unicast,vrf=red}":            100.0,
+		"frr_bgp_peer_message_received_total{afi=ipv4,local_as=64612,peer=192.168.1.3,peer_as=64614,safi=unicast,vrf=red}":            200.0,
+		"frr_bgp_peer_message_received_total{afi=ipv6,local_as=64512,peer=fd00::1,peer_as=64513,safi=unicast,vrf=default}":            29285.0,
+		"frr_bgp_peer_message_received_total{afi=ipv6,local_as=64512,peer=fd00::5,peer_as=64514,safi=unicast,vrf=default}":            0.0,
+		"frr_bgp_peer_message_received_total{afi=ipv6,local_as=64612,peer=fd00::101,peer_as=64613,safi=unicast,vrf=red}":              29285.0,
+		"frr_bgp_peer_message_received_total{afi=ipv6,local_as=64612,peer=fd00::105,peer_as=64614,safi=unicast,vrf=red}":              0.0,
+		"frr_bgp_peer_message_sent_total{afi=ipv4,local_as=64512,peer=192.168.0.2,peer_as=64513,safi=unicast,vrf=default}":            100.0,
+		"frr_bgp_peer_message_sent_total{afi=ipv4,local_as=64512,peer=192.168.0.3,peer_as=64514,safi=unicast,vrf=default}":            0.0,
+		"frr_bgp_peer_message_sent_total{afi=ipv4,local_as=64612,peer=192.168.1.2,peer_as=64613,safi=unicast,vrf=red}":                100.0,
+		"frr_bgp_peer_message_sent_total{afi=ipv4,local_as=64612,peer=192.168.1.3,peer_as=64614,safi=unicast,vrf=red}":                200.0,
+		"frr_bgp_peer_message_sent_total{afi=ipv6,local_as=64512,peer=fd00::1,peer_as=64513,safi=unicast,vrf=default}":                29285.0,
+		"frr_bgp_peer_message_sent_total{afi=ipv6,local_as=64512,peer=fd00::5,peer_as=64514,safi=unicast,vrf=default}":                0.0,
+		"frr_bgp_peer_message_sent_total{afi=ipv6,local_as=64612,peer=fd00::101,peer_as=64613,safi=unicast,vrf=red}":                  29285.0,
+		"frr_bgp_peer_message_sent_total{afi=ipv6,local_as=64612,peer=fd00::105,peer_as=64614,safi=unicast,vrf=red}":                  0.0,
+		"frr_bgp_peer_prefixes_received_count_total{afi=ipv4,local_as=64512,peer=192.168.0.2,peer_as=64513,safi=unicast,vrf=default}": 0.0,
+		"frr_bgp_peer_prefixes_received_count_total{afi=ipv4,local_as=64512,peer=192.168.0.3,peer_as=64514,safi=unicast,vrf=default}": 2.0,
+		"frr_bgp_peer_prefixes_received_count_total{afi=ipv4,local_as=64612,peer=192.168.1.2,peer_as=64613,safi=unicast,vrf=red}":     2.0,
+		"frr_bgp_peer_prefixes_received_count_total{afi=ipv4,local_as=64612,peer=192.168.1.3,peer_as=64614,safi=unicast,vrf=red}":     0.0,
+		"frr_bgp_peer_prefixes_received_count_total{afi=ipv6,local_as=64512,peer=fd00::1,peer_as=64513,safi=unicast,vrf=default}":     1.0,
+		"frr_bgp_peer_prefixes_received_count_total{afi=ipv6,local_as=64512,peer=fd00::5,peer_as=64514,safi=unicast,vrf=default}":     0.0,
+		"frr_bgp_peer_prefixes_received_count_total{afi=ipv6,local_as=64612,peer=fd00::101,peer_as=64613,safi=unicast,vrf=red}":       1.0,
+		"frr_bgp_peer_prefixes_received_count_total{afi=ipv6,local_as=64612,peer=fd00::105,peer_as=64614,safi=unicast,vrf=red}":       0.0,
+		"frr_bgp_peers_count_total{afi=ipv4,local_as=64512,safi=unicast,vrf=default}":                                                 2.0,
+		"frr_bgp_peers_count_total{afi=ipv4,local_as=64612,safi=unicast,vrf=red}":                                                     2.0,
+		"frr_bgp_peers_count_total{afi=ipv6,local_as=64512,safi=unicast,vrf=default}":                                                 2.0,
+		"frr_bgp_peers_count_total{afi=ipv6,local_as=64612,safi=unicast,vrf=red}":                                                     2.0,
+		"frr_bgp_peers_memory_bytes{afi=ipv4,local_as=64512,safi=unicast,vrf=default}":                                                39936.0,
+		"frr_bgp_peers_memory_bytes{afi=ipv4,local_as=64612,safi=unicast,vrf=red}":                                                    39936.0,
+		"frr_bgp_peers_memory_bytes{afi=ipv6,local_as=64512,safi=unicast,vrf=default}":                                                59904.0,
+		"frr_bgp_peers_memory_bytes{afi=ipv6,local_as=64612,safi=unicast,vrf=red}":                                                    59904.0,
+		"frr_bgp_peer_state{afi=ipv4,local_as=64512,peer=192.168.0.2,peer_as=64513,safi=unicast,vrf=default}":                         1.0,
+		"frr_bgp_peer_state{afi=ipv4,local_as=64512,peer=192.168.0.3,peer_as=64514,safi=unicast,vrf=default}":                         0.0,
+		"frr_bgp_peer_state{afi=ipv4,local_as=64612,peer=192.168.1.2,peer_as=64613,safi=unicast,vrf=red}":                             1.0,
+		"frr_bgp_peer_state{afi=ipv4,local_as=64612,peer=192.168.1.3,peer_as=64614,safi=unicast,vrf=red}":                             0.0,
+		"frr_bgp_peer_state{afi=ipv6,local_as=64512,peer=fd00::1,peer_as=64513,safi=unicast,vrf=default}":                             1.0,
+		"frr_bgp_peer_state{afi=ipv6,local_as=64512,peer=fd00::5,peer_as=64514,safi=unicast,vrf=default}":                             0.0,
+		"frr_bgp_peer_state{afi=ipv6,local_as=64612,peer=fd00::101,peer_as=64613,safi=unicast,vrf=red}":                               1.0,
+		"frr_bgp_peer_state{afi=ipv6,local_as=64612,peer=fd00::105,peer_as=64614,safi=unicast,vrf=red}":                               0.0,
+		"frr_bgp_peer_uptime_seconds{afi=ipv4,local_as=64512,peer=192.168.0.2,peer_as=64513,safi=unicast,vrf=default}":                10.0,
+		"frr_bgp_peer_uptime_seconds{afi=ipv4,local_as=64512,peer=192.168.0.3,peer_as=64514,safi=unicast,vrf=default}":                0.0,
+		"frr_bgp_peer_uptime_seconds{afi=ipv4,local_as=64612,peer=192.168.1.2,peer_as=64613,safi=unicast,vrf=red}":                    20.0,
+		"frr_bgp_peer_uptime_seconds{afi=ipv4,local_as=64612,peer=192.168.1.3,peer_as=64614,safi=unicast,vrf=red}":                    0.0,
+		"frr_bgp_peer_uptime_seconds{afi=ipv6,local_as=64512,peer=fd00::1,peer_as=64513,safi=unicast,vrf=default}":                    87873.0,
+		"frr_bgp_peer_uptime_seconds{afi=ipv6,local_as=64512,peer=fd00::5,peer_as=64514,safi=unicast,vrf=default}":                    0.0,
+		"frr_bgp_peer_uptime_seconds{afi=ipv6,local_as=64612,peer=fd00::101,peer_as=64613,safi=unicast,vrf=red}":                      87873.0,
+		"frr_bgp_peer_uptime_seconds{afi=ipv6,local_as=64612,peer=fd00::105,peer_as=64614,safi=unicast,vrf=red}":                      0.0,
+		"frr_bgp_rib_count_total{afi=ipv4,local_as=64512,safi=unicast,vrf=default}":                                                   1.0,
+		"frr_bgp_rib_count_total{afi=ipv4,local_as=64612,safi=unicast,vrf=red}":                                                       0.0,
+		"frr_bgp_rib_count_total{afi=ipv6,local_as=64512,safi=unicast,vrf=default}":                                                   3.0,
+		"frr_bgp_rib_count_total{afi=ipv6,local_as=64612,safi=unicast,vrf=red}":                                                       3.0,
+		"frr_bgp_rib_memory_bytes{afi=ipv4,local_as=64512,safi=unicast,vrf=default}":                                                  64.0,
+		"frr_bgp_rib_memory_bytes{afi=ipv4,local_as=64612,safi=unicast,vrf=red}":                                                      0.0,
+		"frr_bgp_rib_memory_bytes{afi=ipv6,local_as=64512,safi=unicast,vrf=default}":                                                  456.0,
+		"frr_bgp_rib_memory_bytes{afi=ipv6,local_as=64612,safi=unicast,vrf=red}":                                                      456.0,
+	}
+	expectedBgpL2vpnMetrics = map[string]float64{
+		"frr_bgp_l2vpn_evpn_num_arp_nd_count_total{tenantVrf=default,type=L2,vni=172192,vxlanIf=ONTEP1_172192}":       23.000000,
+		"frr_bgp_l2vpn_evpn_num_arp_nd_count_total{tenantVrf=default,type=L2,vni=174374,vxlanIf=ONTEP1_174374}":       0.000000,
+		"frr_bgp_l2vpn_evpn_num_macs_count_total{tenantVrf=default,type=L2,vni=172192,vxlanIf=ONTEP1_172192}":         0.000000,
+		"frr_bgp_l2vpn_evpn_num_macs_count_total{tenantVrf=default,type=L2,vni=174374,vxlanIf=ONTEP1_174374}":         42.000000,
+		"frr_bgp_l2vpn_evpn_num_remote_vteps_count_total{tenantVrf=default,type=L2,vni=172192,vxlanIf=ONTEP1_172192}": 1.000000,
+		"frr_bgp_l2vpn_evpn_num_remote_vteps_count_total{tenantVrf=default,type=L2,vni=174374,vxlanIf=ONTEP1_174374}": 1.000000,
 	}
 )
 
-func TestProcessBGPSummary(t *testing.T) {
-	ch := make(chan prometheus.Metric, 1024)
-	if err := processBGPSummary(ch, bgpSumV4Unicast, "ipv4unicast"); err != nil {
-		t.Errorf("error calling processBGPSummary ipv4unicast: %s", err)
-	}
-	if err := processBGPSummary(ch, bgpSumV6Unicast, "ipv6unicast"); err != nil {
-		t.Errorf("error calling processBGPSummary ipv6unicast: %s", err)
-	}
-	close(ch)
-
-	// Create a map of following format:
-	//   key: metric_name{labelname:labelvalue,...}
-	//   value: metric value
+func prepareMetrics(ch chan prometheus.Metric, t *testing.T) map[string]float64 {
 	gotMetrics := make(map[string]float64)
 
 	for {
@@ -317,9 +341,12 @@ func TestProcessBGPSummary(t *testing.T) {
 
 		gotMetrics[fmt.Sprintf("%s{%s}", metricName, strings.Join(labels, ","))] = value
 	}
+	return gotMetrics
+}
 
+func compareMetrics(t *testing.T, gotMetrics map[string]float64, expectedMetrics map[string]float64) {
 	for metricName, metricVal := range gotMetrics {
-		if expectedMetricVal, ok := expectedBGPMetrics[metricName]; ok {
+		if expectedMetricVal, ok := expectedMetrics[metricName]; ok {
 			if expectedMetricVal != metricVal {
 				t.Errorf("metric %s expected value %v got %v", metricName, expectedMetricVal, metricVal)
 			}
@@ -329,9 +356,34 @@ func TestProcessBGPSummary(t *testing.T) {
 		}
 	}
 
-	for expectedMetricName, expectedMetricVal := range expectedBGPMetrics {
+	for expectedMetricName, expectedMetricVal := range expectedMetrics {
 		if _, ok := gotMetrics[expectedMetricName]; !ok {
 			t.Errorf("missing metric: %s value %v", expectedMetricName, expectedMetricVal)
 		}
 	}
+}
+
+func TestProcessBGPSummary(t *testing.T) {
+	ch := make(chan prometheus.Metric, 1024)
+	if err := processBGPSummary(ch, bgpSumV4Unicast, "ipv4", "unicast"); err != nil {
+		t.Errorf("error calling processBGPSummary ipv4unicast: %s", err)
+	}
+	if err := processBGPSummary(ch, bgpSumV6Unicast, "ipv6", "unicast"); err != nil {
+		t.Errorf("error calling processBGPSummary ipv6unicast: %s", err)
+	}
+	close(ch)
+
+	gotMetrics := prepareMetrics(ch, t)
+	compareMetrics(t, gotMetrics, expectedBGPMetrics)
+}
+
+func TestProcessBgpL2vpnEvpnSummary(t *testing.T) {
+	ch := make(chan prometheus.Metric, 1024)
+	if err := processBgpL2vpnEvpnSummary(ch, evpnVniJson); err != nil {
+		t.Errorf("error calling processBgpL2vpnEvpnSummary: %s", err)
+	}
+	close(ch)
+
+	gotMetrics := prepareMetrics(ch, t)
+	compareMetrics(t, gotMetrics, expectedBgpL2vpnMetrics)
 }

